modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/WalRecordsIteratorAdaptor.java

```diff
@@ -0,0 +1,104 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.persistence.wal;
+
+import java.util.Optional;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.pagemem.wal.WALIterator;
+import org.apache.ignite.internal.pagemem.wal.record.WALRecord;
+import org.apache.ignite.internal.util.GridCloseableIteratorAdapter;
+import org.apache.ignite.lang.IgniteBiTuple;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Iterator over WAL segments. This abstract class provides most functionality for reading records.
+ */
+public abstract class WalRecordsIteratorAdaptor
+    extends GridCloseableIteratorAdapter<IgniteBiTuple<WALPointer, WALRecord>> implements WALIterator {
+    /** */
+    private static final long serialVersionUID = 0L;
+
+    /**
+     * Current record preloaded, to be returned on next()<br> Normally this should be not null because advance() method
+     * should already prepare some value<br>
+     */
+    protected IgniteBiTuple<WALPointer, WALRecord> curRec;
+
+    /**
+     * The exception which can be thrown during reading next record. It holds until the next calling of next record.
+     */
+    private IgniteCheckedException curException;
+
+    /** Logger */
+    @NotNull protected final IgniteLogger log;
+
+    /** Position of last read valid record. */
+    private WALPointer lastRead;
+
+    /**
+     * @param log Logger.
+     */
+    protected WalRecordsIteratorAdaptor(@NotNull final IgniteLogger log) {
+        this.log = log;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteBiTuple<WALPointer, WALRecord> onNext() throws IgniteCheckedException {
+        if (curException != null)
+            throw curException;
+
+        IgniteBiTuple<WALPointer, WALRecord> ret = curRec;
+
+        try {
+            advance();
+        }
+        catch (IgniteCheckedException e) {
+            curException = e;
+        }
+
+        return ret;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected boolean onHasNext() throws IgniteCheckedException {
+        if (curException != null)
+            throw curException;
+
+        return curRec != null;
+    }
+
+    /**
+     * Switches records iterator to the next record. <ul> <li>{@link #curRec} will be updated.</li> </ul>
+     *
+     * {@code advance()} runs a step ahead {@link #next()}
+     *
+     * @throws IgniteCheckedException If failed.
+     */
+    protected abstract void advance() throws IgniteCheckedException;
+
+    /** {@inheritDoc} */
+    @Override public Optional<WALPointer> lastRead() {
+        return Optional.ofNullable(lastRead);
+    }
+
+    /** */
+    protected void lastRead(WALPointer lastRead) {
```

Comment:
This method invokes within `advance` method only. Then the adaptor can be responsible for setting lastRead.

Reply:
Moved the lastRead setting logic into advance method, so that it will be set just before returning the value to caller.